src/ast.ts

```diff
@@ -2045,7 +2073,7 @@ export class FunctionDeclaration extends DeclarationStatement {
   /** Clones this function declaration. */
   clone(): FunctionDeclaration {
     return new FunctionDeclaration(
-      this.name,
+      this.name as IdentifierExpression,
```

Comment:
This can fail silently when compiled to JS. Adding an `assert` above to make sure that the `kind` matches would unify the behavior.

Can you explain why `name` needs to be `Expression`?

Reply:
It needs to be able to hold BindingPatternExpression

Reply:
It's really a `BindingPatternExpression | IdentifierExpression`

Reply:
Hmm, seems odd that this propagates up to `FunctionDeclaration`. Is there maybe a way to separate this?

Reply:
no, because `DeclarationStatement` always holds the property `name`
I recall that @jtenner also thinks this is the only way in a voice chat we had

Reply:
What if the `name` property would become a getter that can be overridden with a more specific type where necessary, say with a private `_name` that is `Expression`? Would that work?

Reply:
That's a good idea, I'll implement that after I have breakfast